### PR TITLE
#5642 fix performance issue when selecting large diff

### LINKED ISSUE #5642: Performance issue when selecting large diff
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Loading a large diff (22.000 changed files) for a commit takes forever. In the current master, loading selecting the diff tab in formbrowse is really slow. It took ~1 second before this commit:
https://github.com/gitextensions/gitextensions/commit/0655fc2e352ba80b62c00df60aa040413ad6c9a6
After this commit it takes over 1 minute. The application freezes and Windows asks to close the not responding application. Because merge commits often have large diff lists, the applications becomes unworkable.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
Select a commit with many changed files (e.g. 20.000). Select the diff tab. This performance hit is also very noticeable on smaller diffs.

To test this in a public repo:
In the linux repository, commit 589109df31843384f410ba09f6b8383464408d1e has 10.000 changed files. Takes ~10 seconds to load on my system. Before it was almost instant. It seems the length of the filenames have impact on the performance as well.

**What is the expected behavior?**
That is loads quickly. Browsing through commits should never hang the application.

**Environment you encounter the issue:**
Commit 0655fc2e352ba80b62c00df60aa040413ad6c9a6

- GitExtensions version: 3.00
- GIT version: 2.19
- OS version: Windows 10

**Did this work in previous version of GitExtensions (which)?**
Yes

### PR DESCRIPTION
**Pull request by NikolayXHD**
Fixes #5642

Changes proposed in this pull request:
- `FileStatusList` does not use `ColumnHeaderAutoResizeStyle.ColumnContent`
anymore, instead column width is determined by visible content only and
is updated on vertical scroll

- Path formatter uses binary search to choose how many characters to
truncate from path

What did I do to test the code and ensure quality:
- Tested on Windows 7
- Observed acceptable performance with linux repository with all `Truncate long filenames` settings

### CODE REVIEW

**Comment by spdr870** [comment]
The performance is improved a lot with this. Thanks.

**Comment by spdr870** [comment]
@NikolayXHD I get an argument exception. Make sure the diff tab is not selected. Open the commit dialog. Close the commit dialog.
Maybe this helps:
image

File: GitUI/UserControls/FileStatusList.cs [hunk 1]
```diff
@@ -493,31 +514,6 @@ int GetLastIndex()
             }
         }
 
-        private static (Image Image, string Text) GetDisplayElements(ListViewItem item, PathFormatter formatter, int itemWidth)
-        {
-            var gitItemStatus = item.Tag<GitItemStatus>();
```
Comment (gerhardol) [inline GitUI/UserControls/FileStatusList.cs]: I got NullReferenceException here, (called from line 1007), not investigating if this still occurs.

**Comment by NikolayXHD** [comment]
> @NikolayXHD I get an argument exception. Make sure the diff tab is not selected. Open the commit dialog. Close the commit dialog.

Fixed, thanks.

File: GitUI/UserControls/FileStatusList.cs (hunk 1 above)
Comment (NikolayXHD) [reply GitUI/UserControls/FileStatusList.cs]: Possbly fixed by e9020e67f3b0c175f216766f68aff1c269438c1e

**Comment by spdr870** [comment]
Tested it, works well.

**Comment by RussKie** [comment]
> Tested it, works well.

If you approve - there is a button 😉

**Pull request merged**